ocaml/xapi/import_raw_vdi.ml

```diff
@@ -98,33 +98,51 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
-              | None, Some sr -> Some sr
-              | None, None -> None
-            in
-            match sr_opt with
-            | Some sr ->
-              debug "Checking whether localhost can see SR: %s" (Ref.string_of sr);
-              if (Importexport.check_sr_availability ~__context sr)
-              then localhost_handler rpc session_id vdi req s
-              else
-                let host = Importexport.find_host_for_sr ~__context sr in
-                let address = Db.Host.get_address ~__context ~self:host in
-                return_302_redirect req s address;
+  try         
+    Server_helpers.exec_with_new_task "VDI.import"
+      (fun __context ->
+         Helpers.call_api_functions ~__context
+           (fun rpc session_id ->
+              let sr_opt = match vdi, sr_of_req ~__context req with
+                | Some vdi, _ -> Some (Db.VDI.get_SR ~__context ~self:vdi)
+                | None, Some sr -> Some sr
+                | None, None -> None
+              in
+              match sr_opt with
+              | Some sr ->
+                debug "Checking whether localhost can see SR: %s" (Ref.string_of sr);
+                if (Importexport.check_sr_availability ~__context sr)
+                then localhost_handler rpc session_id vdi req s
+                else 
+                  let host = Importexport.find_host_for_sr ~__context sr in
+                  let address = Db.Host.get_address ~__context ~self:host in
+                  return_302_redirect req s address;
+                  None
+              | None ->
+                error "Require an SR or VDI to import";
+                let e = Api_errors.(Server_error(vdi_missing,[])) in 
+                let task_id = Xapi_http.task_id_of_req req in
+                begin match task_id with
+                  | Some task_id -> TaskHelper.failed ~__context:(Context.from_forwarded_task task_id) e
+                  | None -> TaskHelper.failed ~__context e
+                end;  
+                Http_svr.headers s (Http.http_400_badrequest ());
                 None
-            | None ->
-              error "Require an SR or VDI to import";
-              TaskHelper.failed ~__context Api_errors.(Server_error(vdi_missing,[]));
-              Http_svr.headers s (Http.http_400_badrequest ());
-              None
-
          )
     )
+  with 
+  | Api_errors.Server_error (a,b) as e ->
+    error "Caught exception in import handler: %s" (ExnHelper.string_of_exn e);
+    let task_id = Xapi_http.task_id_of_req req in
+    begin match task_id with 
+      | Some task_id -> TaskHelper.failed ~__context:(Context.from_forwarded_task task_id) e 
+      | None -> Server_helpers.exec_with_new_task "import" ~task_in_database:true (fun __context -> TaskHelper.failed ~__context e)
+    end;
+    Http_svr.headers s (Http.http_400_badrequest ());
+    None
+  | e ->
+    error "Caught exception in import handler : %s" (Printexc.to_string e);
+    raise e    
```

Comment:
I think it would be good to return an HTTP error and set the task to failed also in this case, where we catch all exceptions. Otherwise the client is never notified, if an error other than `Api_errors.Server_error` happens, and can still hang causing issues like CA-226886 and CA-253489, if I'm not mistaken?